gobo/awesome/sound.lua

```diff
@@ -70,9 +70,30 @@ local function draw_handle(surface, volume)
    cr:fill()
 end
 
-local function draw_lights(surface, state)
+local function darken_pixel(hex_color, start, finish)
+   darker_pixel = string.format("%x", math.floor(tonumber(string.sub(hex_color, start, finish), 16) * 0.4))
+   if (string.len(darker_pixel) < 2) then
+      darker_pixel = "0"..darker_pixel
+   end
+   return darker_pixel
+end
+
+local function darken_color(hex_color)
+   r = darken_pixel(hex_color, 2, 3)
+   g = darken_pixel(hex_color, 4, 5)
+   b = darken_pixel(hex_color, 6, 7)
+   darker_color = "#"..r..g..b
+   return darker_color
+end
+
+local function draw_lights(surface, state, options)
    local cr = cairo.Context(surface)
 
+   local arc_width = options and options.arc_width or 5
+   local arc_mute = options and options.arc_mute or "#ff0000"
+   local arc_fg = options and options.arc_fg or "#00ffff"
+   local arc_fg_darker = darken_color(arc_fg)
+   local arc_bg = options and options.arc_bg or arc_fg_darker or "#006666"
```

Comment:
Instead of passing `options` and processing this every time we run `draw_lights` (including `parse_color` below), you can precompute all that once on `sound.new` and store it in `state`, and then use things like `cr:set_source_rgb(state.arc_mute)`, etc. It will be more efficient and a smaller diff overall!

Reply:
Thanks, i was already thinking of how inneficient it was to put so much stuff in this function but couldn't think of a good way to really fix it.

Reply:
now `draw_lights` receives tables from `state.color_mute`, `state.color_fg`, `state.color_bg`  and `cr:set_source_rgb()` is given as `cr:set_source_rgb(state.color_mute[1], state.color_mute[2], state.color_mute[3])`